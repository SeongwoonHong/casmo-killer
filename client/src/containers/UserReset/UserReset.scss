.Reset {
    > .user-form-box {
        > .user-form {
            > a.user-form-button {
                width: calc(100% - #{$user-box-padding * 2})
            }
        }
    }
}
