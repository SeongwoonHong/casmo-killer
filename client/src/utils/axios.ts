import axios from 'axios';

axios.interceptors.request.use((config) => {
  console.log(config);
  return config;
});

function setTokenToHeader(header, csrfToken) {
  axios.defaults.headers.common[header] = csrfToken;
}

export {
  axios,
  setTokenToHeader,
};
