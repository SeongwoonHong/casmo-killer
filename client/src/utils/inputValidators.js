import inputValidator from 'validator';
import axios from 'axios';

export const trim = (input) => {
  return inputValidator.trim(input);
};

export const validateImg = (img) => {
  return inputValidator.isDataURI(img) || inputValidator.isURL(img);
};

export const validateEmail = async (email) => {

  if (inputValidator.isEmpty(email)) {
    return 'Please enter email address.';
  } else if (/\s/g.test(email)) {
    return 'Email address cannot have any space.';
  } else if (!inputValidator.isEmail(email)) {
    return 'The provided email address is not valid.';
  }

  try {
    const { data } = await axios.get(`/api/auth/verify/email/${email}`);
    if (!data) {
      return 'Internal Server Error.';
    }
    if (data.isDuplicate) {
      return 'The email address is already registered.';
    }
  } catch (error) {
    console.error(error);
    return 'Internal Server Error.';
  }

  return '';

};

export const validateDisplayName = async (displayName) => {

  if (inputValidator.isEmpty(displayName)) {
    return 'Please enter display name.';
  } else if (displayName.length < 4) {
    return 'Display name must be more than 4 characters.';
  } else if (displayName.length > 20) {
    return 'Display name must be less than 20 characters.';
  } else if (/\s/g.test(displayName)) {
    return 'Display name cannot have any space.';
  } else if (!/^[a-zA-Z0-9ㄱ-ㅎ가-힣]{4,20}/.test(displayName)) {
    return 'Display name cannot have special characters.';
  }

  try {
    const { data } = await axios.get(`/api/auth/verify/displayName/${displayName}`);
    if (!data) {
      return 'Internal Server Error.';
    }
    if (data.isDuplicate) {
      return 'The display name is already registered.';
    }
  } catch (error) {
    console.error(error);
    return 'Internal Server Error.';
  }

  return '';

};

export const validatePassword = async (password, confirmPassword = null, isReset = false) => {

  // TODO: password may need more robust validation
  if (inputValidator.isEmpty(password)) {
    return 'Please enter password.';
  } else if (password.length < 6) {
    return 'Password must be more than 6 characters.';
  } else if (password.length > 20) {
    return 'Password must be less than 20 characters.';
  } else if (confirmPassword !== null) {
    if (password !== confirmPassword) {
      return 'Password does not match the confirm password.';
    } else if (password === confirmPassword && !isReset) {
      try {
        const { status } = await axios.post('/api/user/verify/password', { password });
        if (status === 204) {
          return 'New password must be different from current password.';
        }
      } catch (error) {
        console.error(error);
      }
    }

  }

  return '';

};
