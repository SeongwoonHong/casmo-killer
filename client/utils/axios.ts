import axios from 'axios';

const baseURL = 'http://localhost:9000';

axios.defaults.baseURL = baseURL + '/api';

export { axios };
