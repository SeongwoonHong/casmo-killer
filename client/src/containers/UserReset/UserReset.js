import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';

import { validatePassword } from '@sharedUtils/inputValidators';

import FormMessage from '@sharedComponents/FormMessage';

import UserPageContainer from '../../components/UserPageContainer';
import UserInputField from '../../components/UserInputField';

import './UserReset.scss';

class Reset extends Component {

  constructor(props) {

    super(props);

    this.state = {
      isSuccess: false,
      isLoading: true,
      email: '',
      newPassword: '',
      confirmPassword: '',
      message: ''
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onSubmitHandler = this.onSubmitHandler.bind(this);
    this.verifyToken = this.verifyToken.bind(this);

  }

  componentDidMount() {

    const { history, match, setErrorState } = this.props;

    if (match.params.token) {

      // TODO: components that needs token verification can take advantage of HOC that's responsible for token verification
      this.verifyToken(match.params.token);

    } else {

      setErrorState({
        errorTitle: 'The registration link is invalid.',
        errorMsg: 'Please submit your email again to reset the password.'
      });

      history.push('/error');

    }

  }

  onChangeHandler(e) {
    this.setState({ [e.name]: e.value });
  }

  async onSubmitHandler() {

    const { newPassword, confirmPassword } = this.state;

    this.setState({
      isLoading: true,
      message: ''
    });

    const message = await validatePassword(newPassword, confirmPassword, true);

    if (message.length > 0) {
      return this.setState({
        isSuccess: false,
        message
      });
    }

    try {

      const { data } = await axios.put('/api/auth/reset/password', {
        email: this.state.email,
        newPassword: this.state.newPassword
      });

      if (data && data.message) {
        this.setState({
          isSuccess: true,
          message: data.message
        });
      } else {
        this.setState({
          isSuccess: false,
          message: 'Failed to communicate with the server.'
        });
      }

    } catch (error) {

      console.error(error);
      this.setState({
        isSuccess: false,
        message: error.response.data.message
      });
    }

    this.setState({ isLoading: false });

  }

  async verifyToken(token) {

    const { history, setErrorState } = this.props;

    try {

      const { data } = await axios.get(`/api/auth/verify/token/reset/${token}`);

      if (data && data.email) {

        this.setState({
          isLoading: false,
          email: data.email
        });

        this.props.history.replace('/user/reset');

      } else {

        setErrorState({
          errorTitle: 'This link is invalid.',
          errorMsg: 'Failed to communicate with the server.'
        });

        history.push('/error');

      }

    } catch (error) {

      console.error(error);

      setErrorState({
        errorTitle: 'This link is invalid.',
        errorMsg: error.response && error.response.data.message
      });

      history.push('/error');

    }

  }

  render() {

    const {
      isLoading, email, newPassword, confirmPassword, isSuccess, message
    } = this.state;

    return (
      <UserPageContainer
        className="Reset"
        title="Reset Your Password"
        icon="refresh"
        isLoading={ isLoading }
        formTitle="Please enter new password."
        onSubmit={ this.onSubmitHandler }
        button={
          !isSuccess
            ? (
              <button
                type="submit"
                className="user-form-button">
                Submit
              </button>
            )
            : (
              <Link
                to="/user/auth/login"
                className="user-form-button">
                Go to Log In
              </Link>
            )
        }>
        <FormMessage
          message={ message }
          type={ isSuccess ? 'success' : 'error' } />
        <UserInputField
          isVisible={ !isSuccess }
          type="password"
          name="newPassword"
          title="New Password"
          onChange={ this.onChangeHandler }
          value={ newPassword }
          disabled={ email.length === 0 }
          message="Enter new password" />
        <UserInputField
          isVisible={ !isSuccess }
          type="password"
          name="confirmPassword"
          title="Confirm Password"
          onChange={ this.onChangeHandler }
          value={ confirmPassword }
          disabled={ email.length === 0 }
          message="Confirm the new password" />
      </UserPageContainer>
    );

  }

}

export default Reset;
