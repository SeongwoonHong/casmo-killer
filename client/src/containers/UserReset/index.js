import { connect } from 'react-redux';
import * as actions from '@actions';

import UserReset from './UserReset';

function mapStateToProps(state) {
  return {
    auth: state.auth,
    user: state.user
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setErrorState: error => dispatch(actions.setErrorState(error))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(UserReset);
