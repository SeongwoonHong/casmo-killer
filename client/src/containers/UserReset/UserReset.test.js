/* eslint react/jsx-boolean-value: 0 */
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import axios from 'axios';
import { validatePassword } from '@sharedUtils/inputValidators';

import UserReset from './UserReset';


describe('<UserReset />', () => {

  let component;
  let mockGet;
  let mockPut;

  beforeEach(() => {

    component = shallow(
      <UserReset
        setErrorState={ () => {} }
        history={ { push: () => {} } }
        match={ { params: { token: null } } } />
    );

    mockGet = sinon.stub(axios, 'get');
    mockPut = sinon.stub(axios, 'put');

  });

  afterEach(() => {
    mockGet.restore();
    mockPut.restore();
  });

  it('should render the component consistently', () => {
    expect(component).toMatchSnapshot();
  });

  it('should verify the token for registration', async () => {

    const replace = sinon.spy();

    const userData = { data: { email: 'ckboard@gmail.com' } };
    const promise = Promise.resolve(userData);

    mockGet.callsFake(() => promise);

    shallow(
      <UserReset
        history={ { replace } }
        match={ { params: { token: 'abcdefghijk' } } } />
    );

    await promise;

    expect(replace.calledWith('/user/reset')).toEqual(true);

  });

  it('should redirect to error page when the token is not present', () => {

    const setErrorState = sinon.spy();
    const push = sinon.spy();

    shallow(
      <UserReset
        setErrorState={ setErrorState }
        history={ { push } }
        match={ { params: { } } } />
    );

    expect(setErrorState.calledWith({
      errorTitle: 'The registration link is invalid.',
      errorMsg: 'Please submit your email again to reset the password.'
    })).toEqual(true);
    expect(push.calledWith('/error')).toEqual(true);

  });

  it('should check when confirm password is different from new password', async () => {

    const testData = {
      newPassword: 'abcd1234',
      confirmPassword: '1234abcd'
    };

    component.setState(testData);

    component.instance().onSubmitHandler();

    const messageToTest = await validatePassword(testData.newPassword, testData.confirmPassword, true);

    expect(component.state('isSuccess')).toEqual(false);
    expect(component.state('message')).toEqual(messageToTest);

  });

  it('should submit the form to the server and resolve the response to the state', async () => {

    const testData = {
      newPassword: 'abcd1234',
      confirmPassword: 'abcd1234'
    };

    const serverData = { data: { message: 'server response' } };
    const promise = Promise.resolve(serverData);

    mockPut.callsFake(() => promise);

    component.setState(testData);

    component.instance().onSubmitHandler();

    await Promise.resolve();
    await promise;

    expect(component.state('isSuccess')).toEqual(true);
    expect(component.state('message')).toEqual(serverData.data.message);

  });

});
