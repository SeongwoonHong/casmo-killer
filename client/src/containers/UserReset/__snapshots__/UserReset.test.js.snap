// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserReset /> should render the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Reset
    history={
        Object {
            "push": [Function],
          }
    }
    match={
        Object {
            "params": Object {
              "token": null,
            },
          }
    }
    setErrorState={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "button": <button
        className="user-form-button"
        type="submit"
>
        Submit
</button>,
      "children": Array [
        <FormMessage
          message=""
          type="error"
/>,
        <UserInputField
          disabled={true}
          isVisible={true}
          message="Enter new password"
          name="newPassword"
          onChange={[Function]}
          placeholder=""
          title="New Password"
          type="password"
          value=""
/>,
        <UserInputField
          disabled={true}
          isVisible={true}
          message="Confirm the new password"
          name="confirmPassword"
          onChange={[Function]}
          placeholder=""
          title="Confirm Password"
          type="password"
          value=""
/>,
      ],
      "className": "Reset",
      "disabled": false,
      "formMsg": "",
      "formTitle": "Please enter new password.",
      "icon": "refresh",
      "isLoading": true,
      "onSubmit": [Function],
      "title": "Reset Your Password",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": true,
          "isVisible": true,
          "message": "Enter new password",
          "name": "newPassword",
          "onChange": [Function],
          "placeholder": "",
          "title": "New Password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": true,
          "isVisible": true,
          "message": "Confirm the new password",
          "name": "confirmPassword",
          "onChange": [Function],
          "placeholder": "",
          "title": "Confirm Password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "button": <button
          className="user-form-button"
          type="submit"
>
          Submit
</button>,
        "children": Array [
          <FormMessage
            message=""
            type="error"
/>,
          <UserInputField
            disabled={true}
            isVisible={true}
            message="Enter new password"
            name="newPassword"
            onChange={[Function]}
            placeholder=""
            title="New Password"
            type="password"
            value=""
/>,
          <UserInputField
            disabled={true}
            isVisible={true}
            message="Confirm the new password"
            name="confirmPassword"
            onChange={[Function]}
            placeholder=""
            title="Confirm Password"
            type="password"
            value=""
/>,
        ],
        "className": "Reset",
        "disabled": false,
        "formMsg": "",
        "formTitle": "Please enter new password.",
        "icon": "refresh",
        "isLoading": true,
        "onSubmit": [Function],
        "title": "Reset Your Password",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": true,
            "isVisible": true,
            "message": "Enter new password",
            "name": "newPassword",
            "onChange": [Function],
            "placeholder": "",
            "title": "New Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": true,
            "isVisible": true,
            "message": "Confirm the new password",
            "name": "confirmPassword",
            "onChange": [Function],
            "placeholder": "",
            "title": "Confirm Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
